### PR TITLE
CUMULUS-3757: Update granule upsert logic to allow updating collection info

### PR DESCRIPTION
**Pull request by Nnaga1**
**Summary:** Summary of changes

Addresses [CUMULUS-XX: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-XXX)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/db/src/lib/granule.ts [hunk 1]
```diff
@@ -354,3 +360,61 @@ export const getGranulesByGranuleId = async (
     .where({ granule_id: granuleId });
   return records;
 };
+
+/**
+ * Change a granules' PG record and its files' PG record based on collection move
```
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:365]: Nit: formatting.      "Change granules PG record and files" or something like that.

File: packages/db/src/lib/granule.ts [hunk 2]
```diff
@@ -354,3 +360,61 @@ export const getGranulesByGranuleId = async (
...
+ *
+ * @param {Knex | Knex.Transaction} knexOrTransaction - DB client or transaction
```
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:367]: Nit: Should this be either?    I think we probably *don't* want to allow a transaction to create a subtransactio here

File: packages/db/src/lib/granule.ts [hunk 3]
```diff
@@ -354,3 +360,61 @@ export const getGranulesByGranuleId = async (
...
+ * @param {Array<ApiGranuleRecord>} [granules] - updated ApiGranule records
```
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:368]: Are these going to be full ApiGranuleRecords, or is a partial allowed?    This feels like you might want to consider ApiGranule instead, or else enforce a new type that has update fields if they're required.

File: packages/db/src/lib/granule.ts (hunk 1 above)
Comment (Jkovarik) [reply packages/db/src/lib/granule.ts:365]: We need to better define what 'change' means.    Is this an upsert, a bounded modification, etc.

File: packages/db/src/lib/granule.ts [hunk 4]
```diff
@@ -354,3 +360,61 @@ export const getGranulesByGranuleId = async (
...
+        });
+        await trx(granulesTable).where('granule_id', '=', pgGranule.granule_id).update(
+          {
+            collection_cumulus_id: pgGranule.collection_cumulus_id,
+            updated_at: pgGranule.updated_at,
+            status: pgGranule.status,
+            last_update_date_time: pgGranule.last_update_date_time,
+          }
+        );
+        if (granule.files) {
```
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:395]: This doesn't look like it's handling file object set mismatches 'correctly'  - if we're going to expose this business logic as a DB package method, we need to make sure it handles that case correctly or put a logic-bound on making sure the behavior is understood.     

e.g. what does `updateGranulesAndFiles` behavior look like in the case where you have a files mismatch.   

Based on a first read through, it looks like this does a .update to only file records that match (e.g. already exist)... which means if you pass in a file object with *less* or more files, that's ignored.   

Documentation (or documentation via units) and guide-rails are probably needed here if we're writing a method that specific.

File: packages/db/src/lib/granule.ts [hunk 5]
```diff
@@ -354,3 +360,61 @@ export const getGranulesByGranuleId = async (
...
+  knexOrTransaction: Knex | Knex.Transaction,
+  granules: Array<ApiGranuleRecord>
+):Promise<void> => {
+  const {
+    granules: granulesTable,
+    files: filesTable,
+  } = TableNames;
+
+  try {
+    await createRejectableTransaction(knexOrTransaction, async (trx) => {
```
Comment (Jkovarik) [inline packages/db/src/lib/granule.ts:381]: Is it appropriate/intended for the batch update to be all or nothing?  

Nit: This should probably be documented in the header

File: packages/api/endpoints/granules.js [hunk 6]
```diff
@@ -908,6 +910,36 @@ async function getByGranuleId(req, res) {
...
+ * @returns {Promise<Object>} the promise of express response object
+ */
+async function updateGranulesAndFilesCollectionRecords(req, res) {
+  const {
+    knex = await getKnexClient(),
+    esClient = await getEsClient(),
+  } = req.testContext || {};
+  const payload = req.body;
+  try {
+    await updateGranulesAndFiles(knex, payload);
```
Comment (Jkovarik) [inline packages/api/endpoints/granules.js:928]: 🤔  

tl;dr. 

We should probably move the concurrency logic to the API package, handle bounding/configuring concurrency and address validation via something like zod. 

Details/concerns: 

1) We should consider implementing a concurrency or size bound somehow  the API is limited by design to 2 DB connections >by design<, if you throw a large enough request at this endpoint I'd expect it will either timeout or overwhelm the knex connection pool, or both.    That's *probably* not appropriate to the db package (and why we wound up with this sort of logic in the `api/lib` write 'stuff' as an approach. 

2) Users aren't going to just use this for the intended approach for 3757, if it's part of the public API we likely should use zod to validate inputs/schemas/etc before passing a payload to a @cumulus/db method.

File: packages/es-client/indexer.js [hunk 7]
```diff
@@ -115,7 +115,7 @@ async function genericRecordUpdate(esClient, id, doc, index, type, parent) {
  * @param  {string} type     - Elasticsearch type
  * @returns {Promise} Elasticsearch response
  */
-async function updateExistingRecord(esClient, id, doc, index, type) {
+async function updateExistingRecord(esClient, id, doc, index, type, parent = undefined) {
```
Comment (Jkovarik) [inline packages/es-client/indexer.js:118]: Why was this change needed?

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Hey @Nnaga1  - did a once over, the approach looks good 🎉  , but have a few implementation questions for discussion.   Take a look when you have a moment and let's chat!

File: packages/api/endpoints/granules.js (hunk 6 above)
Comment (Jkovarik) [reply packages/api/endpoints/granules.js:928]: I don't think we want to make a bulk update feature like this utilize a single granule call - that's going to force scaling of lambda invocations *per granule*, and is super inefficient.       

There's room for various approaches here, but particularly, we need to put configurations/limits on concurrency around `await Promise.all(granules.map(async (granule) => {` so that we don't overwhelm the limited knex connection pool *or* make an exception for this bulk call and increase it, however that has user monitoring/tuning concerns. 

As an opinionated take:  I'd expect that best looks like having the DB package implement a single granule + all files update and associated tests, with an api lib method that does it for a bunch of granules and has concurrency controls in it.    That allows for solid testing around a single granule update,and decouples concurrency concerns from the DB package.        If the team feels strongly we should support bulk updates as part of the domain of the DB package, we could easily do that same approach entirely within the DB package.
